### PR TITLE
[fix][megatron] Include Fast Hadamard Transform dependency

### PR DESCRIPTION
**Pull request by hershg**
## What

- Add Fast Hadamard Transform to the Megatron extra.
- Pin the tested upstream source revision in project metadata and `uv.lock`.
- Force a CUDA source build compatible with the active PyTorch stack.

## Why

GLM DSA activation rotation imports this package during the first dense forward. Declaring it in the Megatron extra gives the API, engine, and Ray actors one coherent environment and removes both the manual post-lock install and the temporary `uv run --with` overlay.

## Validation

- `uv lock --check`: the 465-package lock is current.
- `git diff --check origin/main..HEAD`: the change is limited to dependency metadata and lockfile entries.
- One B300 / CUDA 13 compatibility receipt: source revision `e7706faf` built for `sm_103`; BF16 forward, inverse reconstruction, backward, and finite gradients passed.
- Two-node standalone SkyRL: canonical `4×4×1` completed 16/16 rollouts, dense log-prob forward, forward/backward, optimizer execution, durable checkpoint, post-step publication, and clean trainer termination.
- Two-node SkyRL–TCLI XID `1048173`: both nodes generated the same lock; `create_model`, publication, a real Tinker 0.25 protobuf sample, sampler checkpoint, unload, and deployment termination passed.
- Two-node SkyRL–TCLI XID `1048176`: exact `4×4×1` completed all 16 requested trajectories, dense log-prob forward, forward/backward, optimizer execution, durable checkpoint, model unload, scheduler release, and successful XM termination.

The full SkyRL CPU suite, code quality, gym, and train test jobs pass. The fork-only GPU workflow fails before tests because its Anyscale credentials are unavailable, and Vercel requires repository authorization.

---

> [!NOTE]
> **Medium Risk**
> Touches Megatron GPU dependency resolution and CUDA extension sourcing; no application code changes, but a bad pin could break installs or Megatron training on Linux.
> 
> **Overview**
> **Adds `fast-hadamard-transform` to the Linux `megatron` optional dependency set** so Megatron/GLM DSA activation paths that import it at runtime are satisfied in one `uv` lock instead of ad-hoc installs or overlays.
> 
> The PR **reverses the previous `override-dependencies` block** that excluded the package (to avoid unpinned PyPI sdist CUDA builds on every install/Ray sync). Resolution is now explicit: **`[tool.uv.sources]`** points x86_64 Python 3.12–3.13 at **custom cu13 + torch 2.11 wheels**, with a **pinned git rev** for other Linux/Python/arch combinations. **Build hooks** add torch/ninja and `FAST_HADAMARD_TRANSFORM_FORCE_BUILD` when a source build is required.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit aaf8aea006f8373773839d4167ca8ad603362ab1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by hershg** [comment]
Downstream B300 validation of the dependency change:

- `fast-hadamard-transform` 1.1.0 pinned to source `e7706faf8d1c3b9f241e36860640ad1dac644ede` built under CUDA 13.0 and produced a native `sm_103` cubin;
- one-node receipt passed BF16 inverse reconstruction, backward, and finite-gradient checks;
- standalone GLM 5.2 completed dense log-prob forward, forward/backward, optimizer, and checkpoint;
- TCLI XIDs 1048304, 1048311, 1048316, and 1048334 completed real forward/backward requests through 16-, 64-, 512-trajectory, and three-step workloads respectively.

The canonical runtime applies the same pyproject/source pin before `uv lock` + frozen sync. Main test suites on this PR are green; the GPU workflow failure is missing Anyscale credentials rather than a code/test failure.

**Comment by hershg** [comment]
Resolved the Linux-scoping comments in 778a7ed: both the Megatron extra and git source are Linux-only, with matching lockfile edges. I tested the suggested `match-runtime = true` change as well, but uv rejects it because fast-hadamard-transform has no static metadata, so runtime matching is impossible for this package. The explicit `torch==2.11.0` build pin therefore remains aligned with this branch’s pinned `torch==2.11.0` CUDA 13 runtime. `uv lock --check` now passes.

File: pyproject.toml [hunk 1]
```diff
@@ -286,9 +287,11 @@ override-dependencies = [
 
 [tool.uv.extra-build-dependencies]
 flash-attn = [{requirement = "torch", match-runtime = true}]
+fast-hadamard-transform = ["torch==2.11.0", "ninja"]
```
[the remark this answers is not in the document]
Comment (erictang000) [reply pyproject.toml:282]: thank you mr. cursor
